server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java

```diff
@@ -334,9 +337,35 @@ private Map<String, AttributeMetadata> configureMetadata(List<AttributeMetadata>
             }
         }
 
+        metadatas.putAll(getUserStorageProviderMetadata(profileMetadata));
```

Comment:
But maybe this will work fine (at least for built-in providers) .

If we want to go that path, I think that we can update LDAP provider to avoid adding attributes, which are available by LDAP mappers (step 1 in `LDAPStorageProvider.decorateUserProfile` ).  It can be probably fine as administrator typically define LDAP mappers for the attributes, which are defined in user profile as these are the attributes, which he is interested about at. There can be corner cases like for instance LDAP would have mapper for the attribute `street` where the attribute `street` is not defined in user-profile configuration. But this is probably fine as why would someone define LDAP mapper for the attribute `street` if not interested about that attribute (AKA. attribute not defined in user profile)?

Not sure if it can be considered use-case that attribute like `street` are available just for the users from `ldap-provider1` when other attributes like `birthdate` are available just for the users from provider `ldap-provider2` etc. Does this use-case make any sense to you?

Reply:
At least the LDAP provider is already working like that where attributes are only added if not already available from the realm UP config, isn't it?

Reply:
Yes, true. We should be hopefully fine here.